### PR TITLE
CPP: Add a query to find incorrectly used exceptions.

### PR DESCRIPTION
**Pull request by ihsinme**
Good day.
in this request I am looking for errors when using exceptions.
I have identified 2 main areas:
1. the possibility of generating an exception in the body of the distructor, without full processing inside.
2. creating an exception object without generating the last one.

Regarding fixes in real software, my proposal is pending, but I plan to continue searching.
https://github.com/assimp/assimp/pull/3954

### CODE REVIEW

File: cpp/ql/src/experimental/Security/CWE/CWE-703/FindIncorrectlyUsedExceptions.ql [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  or
+  fc.getTarget() instanceof Constructor and
+  fc.getTargetType().(Class).getABaseClass+().hasGlobalOrStdName("exception") and
+  not fc.isInMacroExpansion() and
+  not exists(ThrowExpr texp | fc.getEnclosingStmt() = texp.getEnclosingStmt()) and
+  not exists(FunctionCall fctmp | fctmp.getAnArgument() = fc) and
+  not fc instanceof ConstructorDirectInit and
+  not fc.getEnclosingStmt() instanceof DeclStmt and
+  not fc instanceof ConstructorDelegationInit and
+  msg = "This object does not generate an exception."
```
Comment (geoffw0) [inline cpp/ql/src/experimental/Security/CWE/CWE-703/FindIncorrectlyUsedExceptions.ql:37]: This case looks interesting.  Can you give an example (preferably a test case) demonstrating what it is trying to catch?

File: cpp/ql/src/experimental/Security/CWE/CWE-703/FindIncorrectlyUsedExceptions.ql [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+from FunctionCall fc, string msg
+where
+  exists(ThrowExpr texp |
+    texp.getEnclosingFunction() = fc.(DestructorCall).getTarget() and
+    not exists(TryStmt ts |
+      texp.getEnclosingStmt().getParentStmt*() = ts.getStmt() and
+      not ts.getACatchClause().isEmpty()
+    ) and
+    msg = "This distructor contains exeption no wrapped to try..catch blocks"
+  )
```
[the remark this answers is not in the document]
Comment (ihsinme) [reply cpp/ql/src/experimental/Security/CWE/CWE-703/FindIncorrectlyUsedExceptions.ql:27]: if i missed an existing request i apologize. I'll review it again.

File: cpp/ql/src/experimental/Security/CWE/CWE-703/FindIncorrectlyUsedExceptions.ql (hunk 1 above)
Comment (ihsinme) [reply cpp/ql/src/experimental/Security/CWE/CWE-703/FindIncorrectlyUsedExceptions.ql:37]: Actually I am looking for objects of type `std :: exception` that do not throw exceptions.

a basic example can be seen in PR. I will try to make a test case a little later.

**Comment by ihsinme** [comment]
Good day.
Changed description and request a bit.
so as not to disagree with the existing request.

File: cpp/ql/src/experimental/Security/CWE/CWE-703/FindIncorrectlyUsedExceptions.qhelp [hunk 3]
```diff
@@ -7,21 +7,17 @@
...
+<p>Finding for places of incomplete use of objects, type exception.</p>
 <sample src="FindIncorrectlyUsedExceptions.cpp" />
 
 </example>
 <references>
 
 <li>
   CERT CPP Coding Standard:
-  <a href="https://wiki.sei.cmu.edu/confluence/display/cplusplus/DCL57-CPP.+Do+not+let+exceptions+escape+from+destructors+or+deallocation+functions">DCL57-CPP. Do not let exceptions escape from destructors or deallocation functions</a>.
+  <a href="https://wiki.sei.cmu.edu/confluence/display/java/ERR00-J.+Do+not+suppress+or+ignore+checked+exceptions">ERR00-J. Do not suppress or ignore checked exceptions</a>.
```
Comment (geoffw0) [inline cpp/ql/src/experimental/Security/CWE/CWE-703/FindIncorrectlyUsedExceptions.qhelp:20]: This page is about the Java language.  Please would you replace it with an equivalent C/C++ link?

File: cpp/ql/src/experimental/Security/CWE/CWE-703/FindIncorrectlyUsedExceptions.ql [hunk 4]
```diff
@@ -18,12 +17,12 @@ import cpp
...
 where
   exists(ThrowExpr texp |
-    texp.getEnclosingFunction() = fc.(DestructorCall).getTarget() and
+    texp.getEnclosingFunction() = fc.getTarget() and
     not exists(TryStmt ts |
       texp.getEnclosingStmt().getParentStmt*() = ts.getStmt() and
       not ts.getACatchClause().isEmpty()
     ) and
-    msg = "This distructor contains exeption no wrapped to try..catch blocks"
+    msg = "Exception inside this function requires more handling."
```
Comment (geoffw0) [inline cpp/ql/src/experimental/Security/CWE/CWE-703/FindIncorrectlyUsedExceptions.ql:25]: So this is now looking for a throw statement inside a called function, that is not caught/handled inside that function.  This seems like something that normal correct code would be doing fairly frequently, for example a function that's supposed to access a file might reasonably throw an exception if that file doesn't exist.

File: cpp/ql/src/experimental/Security/CWE/CWE-703/FindIncorrectlyUsedExceptions.ql (hunk 1 above)
Comment (geoffw0) [reply cpp/ql/src/experimental/Security/CWE/CWE-703/FindIncorrectlyUsedExceptions.ql:36]: So, broadly, it's looking for cases where a `std::exception` object is created but not as part of a `throw`?  e.g. the author writes
```
std::bad_alloc();
```
when they meant to write:
```
throw std::bad_alloc();
```

**Comment by ihsinme** [comment]
it looks stupid. (
probably the best solution, in a situation where you missed an existing request, is to close the PR and return with a new verified request.

sorry for wasting time.

**Pull request closed without merging**